02cpp1/sec03Pointers.md

```diff
@@ -0,0 +1,364 @@
+
+You also can't make a copy of a unique pointer, as then there would be a conflict over which one should handle the destruction of the object when it goes out of scope. This means that when we want to pass a unique pointer to a function, we cannot pass it by value, because this would involve making a copy. We can, however, pass a unique pointer by reference. 
+```cpp
+#include <memory>
+#include <iostream>
+
+void updatePtrValue(std::unique_ptr<int> &p)
+{
+    *p += 5;
+}
+
+int main()
+{
+    std::unique_ptr<int> p_int = std::make_unique<int>(17);
+    updatePtrValue(p_int);
+    std::cout << *p_int << std::endl;
+
+    return 0;
+}
+```
+This would usually be the preferred way to pass unique pointers to a function. If, for some reason, you don't want to pass by reference, then you will need to use `std::move`. 
+
+Since we can't have multiple unique pointers pointing to the same data, if we want to transfer the ownership of the data to a new unique pointer, we use `std::move` as follows:
+```cpp
+std::unique_ptr<int> p1 = std::make_unique<int>(5);
+std::unique_ptr<int> p2 = std::move(p1);
+```
+- **Avoid doing this if at all possible. After this operation, `p1` will no longer point to valid memory, and will cause a segmentation fault if accessed (i.e. your program will crash).** 
+
+This can also apply to functions if not passing a unique pointer by reference. This can lead to extremely dangerous code as we can see in this example:
+```cpp
+std::unique_ptr<int> updatePtrValue(std::unique_ptr<int> p)
+{
+    *p += 5;
+    return std::move(p);
+}
+
+int main()
+{
+    std::unique_ptr<int> p1 = std::make_unique<int>(17);
+    auto p2 = updatePtrValue2(std::move(p1));
+    p2.swap(p1);
+    std::cout << *p1 << std::endl;
+
+    return 0;
+}
+```
+- In this code, the memory ownership is passed from `p1` to `p`, a unique pointer which is local to the function. In order to avoid the memory being deleted when we leave the function scope, we use `std::move` to move the ownership from `p` to `p2`. At that point, `p2` points to our useful memory, and `p1` is dangling. We use `swap` to move the memory back to `p1`, which leaves `p2` dangling. **If we access `p2` this program will crash. Do not use `std::move` to pass data around, and only use it is you want a new variable to take control of the destruction of that data.** 
+- A good example of using `std::move` would be if another object, perhaps with a broader scope than the existing pointer, needed to take ownership of that data as part of its class. Then if that object outlives the current scope, the data will be maintained for as long as that object lives. **In this case we still need to be careful not to access dangling pointers created by `std::move`**. 
+- We can test whether a unique pointer `p` points to valid memory using `(p)` which returns `true` if it points to valid memory and `false` otherwise.
```

Comment:
`(p)` does not return a bool, it only does so in a statement that expects a bool (e.g. if) where it's implicitly converted. e.g. `if (p) ...` and `bool is_pointing = bool(p)` both work but `bool is_pointing = (p)` does not.